Version 1.5

Gabriel-T-Harris opened the pull request:
Added notes for version 2.

Slight user friendliness improvements.

Moved where the simulation timer is stopped.

Slight cosmetic changes.

@AlexOxorn, if any, what changes did you want to see in version 1?

Gabriel-T-Harris committed:
Notes Changes

Created folder to store notes.

Added notes for Version 2.

Gabriel-T-Harris committed:
User Friendliness

Made things a tad bit more user friendly.

Gabriel-T-Harris committed:
Minor Style Fixes

Corrected a few cosmetic errors.

Gabriel-T-Harris committed:
Changed Timer Stop

Moved when to stop the simulation timer.

Also made some other cosmetic changes.

Gabriel-T-Harris committed:
Minor Wording Change

Gabriel-T-Harris committed:
Added Progress Displaying

While ill advised to actually have progress displaying on due to the
amount that it slows down the program, it has been added.

Gabriel-T-Harris committed:
Improved Directory Supplying

Certain oddities can happen with trailing slashes, so left a warning.

Gabriel-T-Harris commented:
The progress bar should be changed to run in their own thread.

Gabriel-T-Harris committed:
Altered Displaying Progress

Now runs in own thread thereby resulting in it no longer being a
bottleneck. As such is now on by default. Related stuff has also been
updated.

AlexOxorn commented on simulation/Simulation.java:
```diff
@@ -314,51 +314,36 @@ else if (CORE_COUNT < 1)
         final long START_TIME = System.currentTimeMillis(); //simulation start time in milliseconds
-        final long HITS[] = new long[this.FOREST.size()];
-        ArrayList<Deck_Card> current_hand;
+        final int HITS[] = new int[this.FOREST.size()];
```
I feel like `long` makes more sense as it seems reasonable that one would want to run an absurd number of simulations

Gabriel-T-Harris replied:
My reasoning is that the max value of a given index is the number of hands being tested. Thus long feels like a waste of space.

And yes, I would use alias for such in C++.

AlexOxorn commented on simulation/Simulation.java:
```diff
@@ -314,51 +314,36 @@ else if (CORE_COUNT < 1)
-                System.out.printf("\r%3.5f%% of hands drawn.", ++i * PROGRESS_COEFFICIENT);
-            }
+                    System.out.printf("\r%.5f%% of hands drawn.", hands_drawn.floatValue() * PROGRESS_COEFFICIENT);
+                } while (hands_drawn.get() < TEST_HAND_COUNT);
```
standard gratuitous do while blah balh, either you care or you don't.
On a more serious point, consider adding a sleep between prints. We don't need constant printing, and it wastes resources

Gabriel-T-Harris replied:
Yeah I meant to put a sleep.

What is gratuitous about it again?

AlexOxorn replied:
do while loops should be used only when you want to run in the first iteration IN SPITE OF a failing initial loop condition.
For example, looping through a circularly linked list, where start == end.
If the loop condition initially passes anyway, do while(); is gratuitous.
I don't care enough, you do you

Gabriel-T-Harris replied:
I figured it was something more severe then that.

AlexOxorn commented on simulation/Simulation.java:
```diff
@@ -568,23 +553,40 @@ else if (LOWER_THRESHOLD > BLOCKING_QUEUE_SIZE)
                 {
-                    //synchronise transformation of shallow to deep
-                    synchronized (this)
+                    AtomicInteger hands_drawn = new AtomicInteger(START);
```
I feel like having a singular `hands_drawn` might cause some caching bottle necks. Okay for now, but needs testing.
Best way to test is to if raising how many parallel cores used scales the time of the program reasonabley

Gabriel-T-Harris replied:
Well more serious looking is planned for branch Major_Task_Manager_Changes. What caching problem do you envision?

AlexOxorn commented on simulation/Simulation.java:
```diff
@@ -568,23 +553,40 @@ else if (LOWER_THRESHOLD > BLOCKING_QUEUE_SIZE)
-                        System.out.printf("\r%3.5f%% of hands drawn.", (i + 1) * this.PROGRESS_COEFFICIENT);
+                        do
+                        {
+                            System.out.printf("\r%.5f%% of hands drawn.", hands_drawn.floatValue() * this.PROGRESS_COEFFICIENT);
```
Same thing with adding a sleep

Gabriel-T-Harris replied:
Might as well. Even in a sperate thread and nothing else outputting, should still reduce resource use. I imagine it will not lower overall process time though.

Gabriel-T-Harris committed:
Tweaked Progress Display

Made changes to be more memory efficent and less of a bottle neck. Now
running it only is about 5% slower compared to without it.

Gabriel-T-Harris committed:
Update Simulation.java

Added date.

AlexOxorn commented on simulation/Simulation.java:
```diff
@@ -474,6 +525,34 @@ public Task_Manager(ThreadPoolExecutor TASK_OVERSEER, AtomicInteger[] HITS, Arra
+                {
+                    this.HANDS_DRAWN = new AtomicInteger(0);
+
+                    final Thread SEPERATE_PARALLEL_PROGRESS_THREAD = new Thread(() ->
+                    {
+                        do
+                        {
+                            try
+                            {
+                                Thread.sleep(1000); //no need to run the program like every frame
+                            }
+                            catch (InterruptedException ex)
+                            {
+                                ex.printStackTrace();
+                                System.err.println("Ending SEPERATE_PARALLEL_PROGRESS_THREAD.");
+                                break;
+                            }
+
+                            System.out.printf("\r%.5f%% of hands drawn.", this.HANDS_DRAWN.floatValue() * this.PROGRESS_COEFFICIENT);
+                        } while (this.HANDS_DRAWN.get() < TEST_HAND_COUNT);
+                    }, "SEPERATE_PARALLEL_PROGRESS_THREAD");
```
If the body of the lambda is the same, you could probably write a named function, and use a function reference instead of a dublicate lambda, your call

Gabriel-T-Harris replied:
So like I considered that. But could not think of a good way to reuse the whole body to due their difference. Well with out like making a bunch of loser classes just for polymorphism.

Do you have a good way to do it?

AlexOxorn commented on simulation/Simulation.java:
```diff
@@ -553,22 +632,13 @@ else if (LOWER_THRESHOLD > BLOCKING_QUEUE_SIZE)
-                            System.out.printf("\r%.5f%% of hands drawn.", hands_drawn.floatValue() * this.PROGRESS_COEFFICIENT);
-                        } while (hands_drawn.get() < END);
-                    }).start();
+                int i = START; //technically could tweak stuff to use START as i, chose not to bother
```
any particular reason this isn't in the for loop initialization part?

Gabriel-T-Harris replied:
No good reason. I shall change it.

Gabriel-T-Harris committed:
Minor Parameter Use Change

Essentially tweaked some code to reuse an argument. Thereby removing a
variable.

Gabriel-T-Harris committed:
Merge branch 'Version_1_Tweaks' of https://github.com/Gabriel-T-Harris/Hand_Probability_Calculator.git into Version_1_Tweaks

AlexOxorn commented on simulation/Simulation.java:
```diff
@@ -630,15 +630,13 @@ else if (LOWER_THRESHOLD > BLOCKING_QUEUE_SIZE)
              * @param END partition value exclusive
              */
-            protected void run_subroutine(final int START, final int END)
+            protected void run_subroutine(int START, final int END)
```
clever

Gabriel-T-Harris committed:
Correct Oversight

Essentially missed a verbose guard.
